src/test/encoding/check-generated.sh

```diff
@@ -6,93 +6,121 @@ source $CEPH_ROOT/qa/standalone/ceph-helpers.sh
-	run_in_background pids save_stdout "$tmp4" ceph-dencoder type "$type" select_test "$n" copy_ctor dump_json
-	wait_background pids
-
-	if [ $? -ne 0 ]; then
-	    echo "**** $type test $n encode+decode check failed ****"
-	    echo "   ceph-dencoder type $type select_test $n encode decode"
-	    failed=$(($failed + 3))
-	    continue
-	fi
-
-	# nondeterministic classes may dump nondeterministically.  compare
-	# the sorted json output.  this is a weaker test, but is better
-	# than nothing.
-	deterministic=0
-	if ceph-dencoder type "$type" is_deterministic; then
-	    deterministic=1
-	fi
-
-	if [ $deterministic -eq 0 ]; then
-	    echo "  sorting json output for nondeterministic object"
-	    for f in $tmp1 $tmp2 $tmp3 $tmp4; do
-		sort $f | sed 's/,$//' > $f.new
-		mv $f.new $f
-	    done
-	fi
-
-	if ! cmp $tmp1 $tmp2; then
-	    echo "**** $type test $n dump_json check failed ****"
-	    echo "   ceph-dencoder type $type select_test $n dump_json > $tmp1"
-	    echo "   ceph-dencoder type $type select_test $n encode decode dump_json > $tmp2"
-	    diff $tmp1 $tmp2
-	    failed=$(($failed + 1))
-	fi
-
-	if ! cmp $tmp1 $tmp3; then
-	    echo "**** $type test $n copy dump_json check failed ****"
-	    echo "   ceph-dencoder type $type select_test $n dump_json > $tmp1"
-	    echo "   ceph-dencoder type $type select_test $n copy dump_json > $tmp2"
-	    diff $tmp1 $tmp2
-	    failed=$(($failed + 1))
-	fi
-
-	if ! cmp $tmp1 $tmp4; then
-	    echo "**** $type test $n copy_ctor dump_json check failed ****"
-	    echo "   ceph-dencoder type $type select_test $n dump_json > $tmp1"
-	    echo "   ceph-dencoder type $type select_test $n copy_ctor dump_json > $tmp2"
-	    diff $tmp1 $tmp2
-	    failed=$(($failed + 1))
-	fi
-
-	if [ $deterministic -ne 0 ]; then
-	    run_in_background pids ceph-dencoder type "$type" select_test $n encode export "$tmp1"
-	    run_in_background pids ceph-dencoder type "$type" select_test $n encode decode encode export "$tmp2"
-	    wait_background pids
-
-	    if ! cmp $tmp1 $tmp2; then
-		echo "**** $type test $n binary reencode check failed ****"
-		echo "   ceph-dencoder type $type select_test $n encode export $tmp1"
-		echo "   ceph-dencoder type $type select_test $n encode decode encode export $tmp2"
-		diff <(hexdump -C $tmp1) <(hexdump -C $tmp2)
-		failed=$(($failed + 1))
-	    fi
-	fi
-
-	numtests=$(($numtests + 3))
-    done
+    run_in_background pids test_select_type $type $results
+    running_jobs=$(($running_jobs + 1))
+    if [ "$running_jobs" -eq "$max_parallel_jobs" ]; then
+        wait_background pids
+        running_jobs=0
+    fi
 done < <(ceph-dencoder list_types)
+wait_background pids
 
-rm -f $tmp1 $tmp2 $tmp3 $tmp4
+numtests=$(cat $results/* 2>/dev/null | grep "numtests: " | awk '{sum += $2} END {print sum}')
+failed=$(cat $results/* 2>/dev/null | grep "failed: " | awk '{sum += $2} END {print sum}')
+rm -rf $results
```

Comment:
We should not remove that temporary directory if we have failed jobs.

Reply:
This is the count of test results.  The failure details will be printed in the test loop. So the tmp dir can be remove.